src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/DefaultRazorProjectFileSystem.cs

```diff
@@ -23,67 +40,93 @@ public override IEnumerable<RazorProjectItem> EnumerateItems(string basePath)
+            return ThrowHelper.ThrowInvalidOperationException<RazorProjectItem>($"The file '{absolutePath}' is not a descendent of the base path '{absoluteBasePath}'.");
         }
 
-        var relativePhysicalPath = file.FullName.Substring(absoluteBasePath.Length + 1); // Include leading separator
+        return CreateItem(absolutePath, fileKind, DefaultBasePath, absoluteBasePath);
+    }
+
+    private static DefaultRazorProjectItem CreateItem(string path, string? fileKind, string basePath, string absoluteBasePath)
+    {
+        var physicalPath = Path.GetFullPath(path);
+        var relativePhysicalPath = physicalPath[(absoluteBasePath.Length + 1)..]; // Don't include leading separator
+
         var filePath = "/" + relativePhysicalPath.Replace(Path.DirectorySeparatorChar, '/');
 
-        return new DefaultRazorProjectItem("/", filePath, relativePhysicalPath, fileKind, new FileInfo(absolutePath), cssScope: null);
+        return new DefaultRazorProjectItem(basePath, filePath, physicalPath, relativePhysicalPath, fileKind, cssScope: null);
     }
 
     protected override string NormalizeAndEnsureValidPath(string path)
     {
+        // PERF: If we're asked to normalize "/", there's no need to compare and manipulate strings to
+        // ultimately return the value of Root.
+        if (path == DefaultBasePath)
+        {
+            return Root;
+        }
+
         ArgHelper.ThrowIfNullOrEmpty(path);
 
-        var absolutePath = path.Replace('\\', '/');
+        var normalizedPath = path.Replace('\\', '/');
 
-        // Check if the given path is an absolute path. It is absolute if,
-        // 1. It starts with Root or
-        // 2. It is a network share path and starts with a '//'. Eg. //servername/some/network/folder
-        if (!absolutePath.StartsWith(Root, StringComparison.OrdinalIgnoreCase) &&
-            !absolutePath.StartsWith("//", StringComparison.OrdinalIgnoreCase))
+        // Check if the given path is an absolute path. It is absolute if...
+        //
+        // 1. It is a network share path and starts with a '//' (e.g. //server/some/network/folder) or...
+        // 2. It starts with Root
+        if (normalizedPath is ['/', '/', ..] ||
+            normalizedPath.StartsWith(Root, StringComparison.OrdinalIgnoreCase))
         {
-            // This is not an absolute path. Strip the leading slash if any and combine it with Root.
-            if (path[0] == '/' || path[0] == '\\')
-            {
-                path = path.Substring(1);
-            }
+            return normalizedPath;
+        }
+
+        // This is not an absolute path, so we combine it with Root.
 
-            // Instead of `C:filename.ext`, we want `C:/filename.ext`.
-            absolutePath = Root.EndsWith(':') && !path.IsNullOrEmpty()
-                ? Root + "/" + path
-                : Path.Combine(Root, path);
+        using var builder = new MemoryBuilder<char>(initialCapacity: Root.Length + normalizedPath.Length + 1);
```

Comment:
It feels like it would be straightforward to compute the exact final length and use `string.Create` but perhaps that's an overkill.

Reply:
I actually started with `StringExtensions.CreateString(...)`, but changed it to use our `MemoryBuilder<char>` because it seemed less straightforward. :smile: The problem is one of "capacity" for `MemoryBuilder<char>` vs. "length" for `CreateString(...)`. We actually don't know the length yet, since we have to determine whether we need to add a '/' or not. So, the code has to do a little more work upfront to use `CreateString(...)`. However, using `MemoryBuilder<char>` *does* end up copying the characters twice, so I'll go ahead and switch it back to `CreateString()`.